torch/csrc/jit/passes/autocast.cpp

```diff
@@ -0,0 +1,456 @@
+            use.user->s(attr::name) == "fast_dtype") {
+          // Search for `prim::SetAttr[name="fast_dtype"]`
+          auto ret = constant_as<c10::ScalarType>(use.user->input(1));
+          TORCH_CHECK(
+              ret.has_value() && ret.value() != c10::ScalarType::Undefined,
+              "Autocast dtype argument must be a constant and defined");
+          dtype = ret.value();
+        }
+      }
+      TORCH_CHECK(enabled.has_value(), "Autocast missing _enabled attribute");
+      TORCH_CHECK(
+          dtype != c10::ScalarType::Undefined,
+          "Autocast missing fast_dtype attribute");
+      TORCH_CHECK(!device.empty(), "Autocast missing device attribute");
+      if (device == "cuda") {
+        scope.context.enabled = enabled.value();
+        scope.context.scalar_type = dtype;
+      } else if (device == "cpu") {
+        scope.context.cpu_enabled = enabled.value();
+        scope.context.cpu_scalar_type = dtype;
+      } else {
+        TORCH_INTERNAL_ASSERT(
+            false, "unrecognized device for autocast pass: ", device);
+      }
+      return scope;
+    } else {
+      // We only support simple and static autocast expressions. For example,
+      // the following should report an error (since the autocast would not
+      // work as expected)
+      //
+      //    autocast_on = autocast(enabled=True)
+      //    autocast_off = autocast(enabled=False)
+      //    with autocast_on if condition else autocast_off:
+      //        ...
+      //
+      // TODO: better error message
+      //
+      AT_ERROR("Unsupported autocast syntax");
+    }
+  }
+
+  // Not an autocast...
+  return c10::nullopt;
+}
+
+void castTensorInputs(
+    Node* node,
+    Symbol cast_op,
+    const AutocastContext& context) {
+  if (!context) {
+    return;
+  }
+
+  const auto graph = node->owningGraph();
+
+  std::unordered_set<Value*> casted_inputs;
+  for (auto input : node->inputs()) {
+    // TODO: update cast_op signature to take dynamic context flags
+    auto input_tensor_type = input->type()->cast<TensorType>();
+    if (input_tensor_type && input->node()->kind() != cast_op) {
+      casted_inputs.insert(input);
+    }
+  }
+
+  WithInsertPoint insert_point(node);
+
+  for (auto input : casted_inputs) {
+    if (cast_op == aten::_autocast_to_full_precision) {
+      const auto new_input = graph->insert(
+          cast_op,
+          {input,
+           graph->insertConstant(IValue(context.enabled)),
+           graph->insertConstant(IValue(context.cpu_enabled))});
+      node->replaceInputWith(input, new_input);
+    } else if (cast_op == aten::_autocast_to_reduced_precision) {
+      const auto new_input = graph->insert(
+          cast_op,
+          {input,
+           graph->insertConstant(IValue(context.enabled)),
+           graph->insertConstant(IValue(context.cpu_enabled)),
+           graph->insertConstant(IValue(context.scalar_type)),
+           graph->insertConstant(IValue(context.cpu_scalar_type))});
```

Comment:
do we not need `gpu_scalar_type` just curious ? if there is any chance of gpu bfloat16 casting should we just include this now to make our lives easier ?

Reply:
Sorry for the confusing names.

`context.scalar_type` is supposed to be `context.gpu_scalar_type`, I'm omitting the `gpu` tag there to match eager amp APIs: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/autocast_mode.h#L6-L18

Maybe that's a bad idea. I'll have them properly named here.